tests/Microsoft.Identity.Test.Unit.net45/Integration/DeviceCodeFlowIntegrationTest.cs

```diff
@@ -0,0 +1,169 @@
+{
+    // Important: do not install a NuGet package with the Chrome driver as it is a security risk.
+    // Instead, install the Chrome driver on the test machine
+
+    // Note: these tests require permission to a KeyVault Microsoft account; 
+    // Please ignore them if you are not a Microsoft FTE, they will run as part of the CI build
+    [TestClass]
+    public class DeviceCodeFlow
+    {
+        private static readonly string[] Scopes = { "User.Read" };
+
+        public TestContext TestContext { get; set; }
+
+        [ClassInitialize]
+        public static void ClassInitialize(TestContext context)
+        {
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
+        }
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            TestCommon.ResetState();
+        }
+
+
+        [TestMethod]
+        [Timeout(60 * 1000)] // 1 min timeout
+        public async Task DeviceCodeFlowTestAsync()
+        {
+            Debug.WriteLine("Fetching user from lab");
+            LabResponse labResponse = LabUserHelper.GetDefaultUser();
+
+            Debug.WriteLine("Calling AcquireTokenWithDeviceCodeAsync");
+            PublicClientApplication pca = new PublicClientApplication(labResponse.AppId);
+            var result = await pca.AcquireTokenWithDeviceCodeAsync(Scopes, deviceCodeResult =>
+            {
+                RunAutomatedDeviceCodeFlow(deviceCodeResult, labResponse.User);
+
+                return Task.FromResult(0);
+            }).ConfigureAwait(false);
+
+            Assert.IsNotNull(result);
+            Assert.IsTrue(!string.IsNullOrEmpty(result.AccessToken));
+        }
+
+        private void RunAutomatedDeviceCodeFlow(DeviceCodeResult deviceCodeResult, LabUser user)
+        {
+            IWebDriver driver = null;
+            try
+            {
+                driver = InitDriver();
+
+                Debug.WriteLine("Navigating and enterign the code");
+
+                driver.Navigate().GoToUrl(deviceCodeResult.VerificationUrl);
+                driver.FindElement(By.Id("code")).SendKeys(deviceCodeResult.UserCode);
+
+                IWebElement continueBtn = WaitForElementToBeEnabled(driver, By.Id("continueBtn"));
+                continueBtn?.Click();
+
+                Debug.WriteLine("Logging in");
+                PerformLogin(driver, user);
+
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine("Browser automation failed " + ex);
+#if DESKTOP // Can't attach a file on netcore because mstest doesn't support it
+                Screenshot ss = ((ITakesScreenshot)driver).GetScreenshot();
+                string failurePicturePath = Path.Combine(TestContext.ResultsDirectory, "failure.png");
+                ss.SaveAsFile(failurePicturePath, ScreenshotImageFormat.Png);
+                TestContext.AddResultFile(failurePicturePath);
+
+                Debug.WriteLine("Failing because of " + ex);
+#endif
```

Comment:
ok. This should be baked into the library as we have other places where we might want to use this (though not something we generally do in our integration tests)

Reply:
Our other tests use Appium, so the actual logic is different. Will generalise these helpers when we write more Selenium tests.

Reply:
Should really be separated already now. Next time it will probably not be you.

Reply:
I will have proposed more abstraction in the subsequent PR https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/718